libcontainer/cgroups/fs/memory.go

```diff
@@ -55,9 +58,50 @@ func (s *MemoryGroup) Apply(d *cgroupData) (err error) {
 }
 
 func (s *MemoryGroup) SetKernelMemory(path string, cgroup *configs.Cgroup) error {
-	// This has to be done separately because it has special constraints (it
-	// can't be done after there are processes attached to the cgroup).
-	if cgroup.Resources.KernelMemory > 0 {
+	// This has to be done separately because it has special
+	// constraints (it can only be initialized before setting up a
+	// hierarchy or adding a task to the cgroups. However, if
+	// sucessfully initialized, it can be updated anytime afterwards)
+	if cgroup.Resources.KernelMemory != 0 {
+		kmemInitialized := false
+		// Is kmem.limit_in_bytes already set?
+		if b, err := ioutil.ReadFile(filepath.Join(path, "memory.kmem.limit_in_bytes")); err != nil {
+			return err
+		} else {
+			kmemValue, err := strconv.ParseUint(strings.TrimSpace(string(b)), 10, 64)
+			if err != nil {
+				return err
+			}
+			switch system.GetLongBit() {
+			case 32:
+				kmemInitialized = uint32(kmemValue) != uint32(math.MaxUint32)
+			case 64:
+				kmemInitialized = kmemValue != uint64(math.MaxUint64)
+			}
+		}
+
+		if !kmemInitialized {
```

Comment:
> On my system the memory cgroups have use_hierachy = 0 by default (ubuntu 14.04). Also a runc spec doesn't specify a value for kernel memory which mean it'll be uninitialized by default.

Fair enough, I didn't know that (it doesn't happen on any of my systems).

> And like in the bats test I wrote, when specifying a new cgroupath for the container, they may not have any tasks inside if it's brand new.

Right, but the issue isn't when creating a container (the old code would work in that case). My question is when would we be updating a container's limits without it having processes inside it (isn't that the point of the PR)? The reason the code is written the way it is now is so that Docker could update containers (before the `containerd` split) -- and that worked perfectly fine.